public/js/common.js

```diff
@@ -3,6 +3,41 @@
  */
 const LRR = {};
 
+// This class is used to wrap URLs that point into the app, including
+// API endpoints and browser targets. It reads the configured base URL
+// from the template, then prepends it to the provided URL in the
+// toString() method.
+
+// Every operation involving an app-internal URL should be wrapped in
+// this class before use. It can be passed to any API that expects a
+// URL, and can be wrapped around another instance of itself without a
+// change.
+LRR.apiURL = class {
+    constructor(load_url) {
+        // accept instances of self as well, to make wrapping
+        // idempotent
+        if (load_url instanceof LRR.apiURL) {
+            load_url = load_url.load_url;
+        }
+        this.load_url = load_url;
+        if (!this.load_url.startsWith("/")) {
+            console.trace("passed non-absolute URL to apiURL");
+            this.load_url = "/" + this.load_url;
+        }
+    }
+
+    toString() {
+        // in the default case, this will be empty string and will
+        // leave the load URL unchanged
+        let base_url = window.LRR_BASEURL;
```

Comment:
Is there a way we could do this with an additional header from the mojo/server side instead of requiring a property to be set in every template?  

There's already a lot of boilerplate in all the templates so it's not really a big deal, but I feel it'd be less code.

Reply:
There's no way to read an arbitrary header from the same-page request using Javascript. However, it turns out you can set a cookie and then read that.

I do like this better than putting it in template code. That's just another thing to maybe forget later.

Reply:
huh, I thought headers were accessible somehow but I'll admit I never tried to do it 🙇 The cookie approach sounds good to me.